casatables/src/lib.rs

```diff
@@ -1773,6 +2209,306 @@ impl Drop for TableRow {
+                )
+            };
+
+            if rv != 0 {
+                return self.exc_info.as_err();
+            }
+        } else if T::DATA_TYPE == glue::GlueDataType::TpArrayString {
+            let glue_strings = T::casatables_stringvec_pass_through_out(value);
+
+            let rv = unsafe {
+                glue::tablerec_put_field(
+                    self.handle,
+                    &cfield_name,
+                    T::DATA_TYPE,
+                    shape.len() as u64,
+                    shape.as_ptr(),
+                    glue_strings.as_ptr() as _,
+                    &mut self.exc_info,
+                )
+            };
+
+            if rv != 0 {
+                return self.exc_info.as_err();
+            }
+        } else {
+            let rv = unsafe {
+                glue::tablerec_put_field(
+                    self.handle,
+                    &cfield_name,
+                    T::DATA_TYPE,
+                    shape.len() as u64,
+                    shape.as_ptr(),
+                    value.casatables_as_buf() as _,
+                    &mut self.exc_info,
+                )
+            };
+
+            if rv != 0 {
+                return self.exc_info.as_err();
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl PartialEq for TableRecord {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe { glue::tablerec_eq(self.handle, other.handle) }
+    }
+}
+
+impl Eq for TableRecord {}
+
+impl Debug for TableRecord {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // This is to avoid mutably borrowing self
+        let mut exc_info = unsafe { std::mem::zeroed::<glue::ExcInfo>() };
+        write!(f, "TableRecord {{ ").unwrap();
+        unsafe { invoke_tablerec_get_keyword_repr(self.handle, &mut exc_info, |name, type_, repr| {
+            write!(f, "{}:{} = {}, ", name, type_, repr).unwrap();
+        }) };
+        write!(f, " }}")
+    }
+}
+
+impl CasaDataType for TableRecord {
+    const DATA_TYPE: glue::GlueDataType = glue::GlueDataType::TpRecord;
+
+    fn casatables_alloc(_shape: &[u64]) -> Result<Self, Error> {
+        TableRecord::new()
+    }
+
+    fn casatables_as_buf(&self) -> *const () {
+        self.handle as _
+    }
+
+    fn casatables_as_mut_buf(&mut self) -> *mut () {
+        self.handle as _
+    }
+
+    /// A hack that lets us properly special-case table records.
+    #[doc(hidden)]
+    fn casatables_tablerec_pass_through(r: TableRecord) -> Self {
+        r
+    }
+
+    /// A hack that lets us properly special-case table records
+    #[doc(hidden)]
+    fn casatables_tablerec_pass_through_out(s: &Self) -> TableRecord {
+        s.to_owned()
+    }
+}
+
+impl Drop for TableRecord {
+    fn drop(&mut self) {
+        // TODO: we want to free standalone tablerecords, but not those which
+        // are associated with a table, because they are freed automatically
+        // with the table. Not sure the best way to do this.
```

Comment:
Yikes, does this mean that there's a potential for a double-free to happen with this drop function? It would feel a bit fragile but we could potentially add some kind of `is_owned` flag to the `TableRecord` struct and check that in the drop implementation. Or maybe some way to use a C++ refcounted pointer type on that side ...?